sympy/sets/ordinals.py

```diff
@@ -0,0 +1,176 @@
+from sympy.core import Basic
+import operator
+
+class Ordinal(Basic):
+    """
+    Represents ordinals in Cantor normal form.
+
+    Internally, this class is just a list of lists (known
+    as "terms").
+    In single terms [a, b], a represents power to the ordinal and b represents multiplying factor
+    Examples
+    ========
+    >>> from sympy.sets import Ordinal, Ordinals
+    >>> Ordinals.w  # doctest: +SKIP
+    w
+    >>> a = Ordinals.w
+    >>> a.is_limit_ordinal
+    True
+    >>> Ordinal([[3,2]])  # doctest: +SKIP
+    {w**3}*2
+    >>> Ordinal([[5,1],[3,2]])  # doctest: +SKIP
+    w**5 + {w**3}*2
+    """
+    def __new__(cls, terms):
+        obj = super(Ordinal, cls).__new__(cls, *terms)
+        powers = [i[0] for i in obj.args]
+        if not all(earlier >= later for earlier, later in zip(powers, powers[1:])):
+            raise ValueError("powers must be in decreasing order")
+        return obj
+
+    @property
+    def is_successor_ordinal(self):
+        return (self.args[-1][0] == 0)
+
+    @property
+    def is_limit_ordinal(self):
+        return not self.is_successor_ordinal
+
+    def _insert_index(self, a, x, lo=0, hi=None):
+        """Return the index where to insert item x in list a,
+        assuming a is sorted in descending order
+
+        Essentially, the function returns number of elements in a which are >= than x.
+        """
+        if lo < 0:
+            raise ValueError('lo must be non-negative')
+        if hi is None:
+            hi = len(a)
+        while lo < hi:
+            mid = (lo+hi)//2
+            if x > a[mid]: hi = mid
+            else: lo = mid+1
+        return lo
+
+    def _compare(self, other, index):
+        value_self = [i[index] for i in self.args]
+        value_other = [i[index] for i in other.args]
+        if value_other == value_self:
+            return -1
+        else:
+            return map(operator.eq, value_self, value_other).index(False)
+
+    def __eq__(self, other):
+        if isinstance(other, Ordinal):
+            return self.args == other.args
+        else:
+            return False
+
+    def __lt__(self, other):
+        if isinstance(other, int):
```

Comment:
I would not define automatic comparison of ordinals and ordinary (cardinal) numbers. (There could be a method for converting an integer to a finite ordinal, but that should be used explicitly.) I would return `NotImplemented` when `other` is not an ordinal.